src/hashutils.rs

```diff
@@ -1,27 +1,31 @@
 
-use ring::digest::{ Algorithm, Context, digest, Digest };
+use ring::digest::{ Algorithm, Context, Digest };
 
 /// The sole purpose of this trait is to extend the standard
 /// `ring::algo::Algorithm` type with a couple utility functions.
 pub trait HashUtils {
 
-    /// Compute the hash the given byte array
-    fn hash_bytes(&'static self, bytes: &AsRef<[u8]>) -> Digest;
+    /// Compute the hash of the given leaf
+    fn hash_leaf(&'static self, bytes: &AsRef<[u8]>) -> Digest;
 
     /// Compute the hash of the concatenation of `left` and `right`.
     // XXX: This is overly generic temporarily to make refactoring easier.
     // TODO: Give `left` and `right` type &Digest.
-    fn combine_hashes(&'static self, left: &AsRef<[u8]>, right: &AsRef<[u8]>) -> Digest;
+    fn hash_nodes(&'static self, left: &AsRef<[u8]>, right: &AsRef<[u8]>) -> Digest;
 }
 
 impl HashUtils for Algorithm {
 
-    fn hash_bytes(&'static self, bytes: &AsRef<[u8]>) -> Digest {
-        digest(self, bytes.as_ref())
+    fn hash_leaf(&'static self, leaf: &AsRef<[u8]>) -> Digest {
+        let mut ctx = Context::new(self);
+        ctx.update(&[0x00]);
+        ctx.update(leaf.as_ref());
+        ctx.finish()
     }
 
-    fn combine_hashes(&'static self, left: &AsRef<[u8]>, right: &AsRef<[u8]>) -> Digest {
+    fn hash_nodes(&'static self, left: &AsRef<[u8]>, right: &AsRef<[u8]>) -> Digest {
         let mut ctx = Context::new(self);
+        ctx.update(&[0x01]);
         ctx.update(left.as_ref());
```

Comment:
It's great that we're respecting [CT hashing values](https://tools.ietf.org/html/rfc6962#page-4). Should we also support the empty scenarios so that `merkle` can be used as a dependency for a CT library? Currently, empty vec will return `None` but we might want to return the hash of an empty string instead.